File: native/platform/macos/swift/App/CloudConfig.swift
import Foundation
import GRPCNIOTransportHTTP2

/// Single source of truth for the Pivox cloud gRPC endpoint and
/// transport security. Both `ChatClient` and `OrgsClient` resolve
/// through here so they can't drift.
///
/// Resolution order:
///   1. `PIVOX_GRPC_HOST` env var, e.g. "localhost:50051" (no scheme).
///      `PIVOX_GRPC_PLAINTEXT=1` disables TLS — useful for local dev
///      pointing at a plaintext server on a non-443 port.
///   2. Default: `pivox.ngrok.app:443` over TLS. Matches the public
///      tunnel used by the OAuth broker, so a single environment
///      override (or no override at all) keeps gRPC + REST + auth
///      pointing at the same backend.
///
/// Local-dev workflow: export `PIVOX_GRPC_HOST=localhost:50051` and
/// `PIVOX_GRPC_PLAINTEXT=1` in `.envrc`. Production builds inherit
/// the ngrok default.
enum CloudConfig {
  /// "host:port" string, no scheme.
  static let grpcEndpoint: String = {
    if let override = ProcessInfo.processInfo.environment["PIVOX_GRPC_HOST"],
       !override.isEmpty {
      return override
    }
    return "pivox.ngrok.app:443"
  }()

  /// Whether to negotiate TLS on the gRPC channel. ngrok terminates
  /// TLS on :443; local dev typically runs plaintext on :50051.
  static let usePlaintext: Bool = {
    ProcessInfo.processInfo.environment["PIVOX_GRPC_PLAINTEXT"] == "1"
  }()

  /// Resolved transport security for grpc-swift-2's
  /// `HTTP2ClientTransport.Posix`. Use this directly when constructing
  /// a client transport.
  static var transportSecurity: HTTP2ClientTransport.Posix.TransportSecurity {
    usePlaintext ? .plaintext : .tls
  }

  /// Parses `grpcEndpoint` into (host, port). Throws on malformed
  /// values rather than silently picking a default — a typo here
  /// would otherwise misroute every RPC.
  static func parsedEndpoint() throws -> (host: String, port: Int) {
    let parts = grpcEndpoint.split(separator: ":", maxSplits: 1).map(String.init)
    guard parts.count == 2, let port = Int(parts[1]) else {
      throw CloudConfigError.invalidEndpoint(grpcEndpoint)
    }
    return (parts[0], port)
  }
}

enum CloudConfigError: Error, CustomStringConvertible {
  case invalidEndpoint(String)

  var description: String {
    switch self {
    case .invalidEndpoint(let s): return "invalid PIVOX_GRPC_HOST: \(s)"
    }
  }
}

File: dotnet/Pivox.Shared/CloudConfig.cs
namespace Pivox.Shared;

/// <summary>
/// Single source of truth for the Pivox cloud gRPC endpoint.
///
/// Resolution order:
///   1. <c>PIVOX_GRPC_HOST</c> env var, e.g. "pivox.ngrok.app:443"
///      (host:port, NO scheme).
///   2. Default: <c>pivox.ngrok.app:443</c>. Matches the public
///      tunnel the OAuth broker uses, so gRPC + REST + auth all
///      point at the same backend.
///
/// Transport: TLS only. There is no plaintext deployment mode for
/// the dotnet stack — broker callbacks carry id tokens, gRPC carries
/// Firebase JWTs, neither can ever leave the device unencrypted.
///
/// Divergence from the SwiftUI <c>CloudConfig.swift</c>: the Swift
/// side retains a <c>PIVOX_GRPC_PLAINTEXT</c> escape hatch for local
/// development against a no-TLS gRPC server. The dotnet side
/// deliberately doesn't — that path was never exercised here, and
/// removing it shrinks the auth surface area. SwiftUI is on the
/// retirement track; the parity was a transitional convenience that
/// has expired.
/// </summary>
public static class CloudConfig
{
    /// <summary>"host:port" string, no scheme.</summary>
    public static string GrpcEndpoint
    {
        get
        {
            var env = Environment.GetEnvironmentVariable("PIVOX_GRPC_HOST");
            return !string.IsNullOrEmpty(env) ? env : "pivox.ngrok.app:443";
        }
    }

    /// <summary>
    /// Composes the endpoint into an https:// <see cref="Uri"/> for
    /// <c>GrpcChannel.ForAddress</c>. Always TLS.
    /// </summary>
    public static Uri GrpcUri
    {
        get
        {
            var (host, port) = ParsedEndpoint();
            return new Uri($"https://{host}:{port}");
        }
    }

    /// <summary>
    /// Base URL for pivox-cloud's REST surface: SSO provider resolution,
    /// the OAuth/OIDC broker, etc. Derived from <see cref="GrpcEndpoint"/>'s
    /// host so a single <c>PIVOX_GRPC_HOST</c> override flips both gRPC
    /// and REST. Always HTTPS.
    /// </summary>
    public static string BrokerBaseUrl
    {
        get
        {
            var (host, _) = ParsedEndpoint();
            return $"https://{host}";
        }
    }

    /// <summary>
    /// Parses <see cref="GrpcEndpoint"/> into (host, port). Throws on
    /// malformed values rather than silently picking a default — a
    /// typo here would otherwise misroute every RPC.
    /// </summary>
    public static (string Host, int Port) ParsedEndpoint()
    {
        var parts = GrpcEndpoint.Split(':', 2);
        if (parts.Length != 2 || !int.TryParse(parts[1], out var port))
        {
            throw new InvalidOperationException(
                $"Invalid PIVOX_GRPC_HOST: '{GrpcEndpoint}' (expected host:port).");
        }
        return (parts[0], port);
    }
}
